### ISSUE TITLE
Build custom Error Details message for REST API

### ISSUE THREAD

**Issue by cazacmarin** [hzi-braunschweig/SORMAS-Project#8934]
### Situation Description
Currently, Sormas Rest API is handling a small range of errors which can be reported to the front(or service which is interrogating REST endpoints).
If appear a scenario where a new specific error needs to be added, it is done in if/else block and those blocks will increase is size.
Moreover, they contain only a specific text message. 
### Feature Description
- To have a custom builder for errors which will provide more details about the error message.
- To have specific group of error codes. Error: response status is 401, 403, 406, 409, etc.
- To have a defined constants in case if a common message can appear. On front side can be applied internationalization based on error message ID. Otherwise, default message string in english language can be displayed.
Expected output:
{
  "timestamp": "2022-04-22T16:03:40.602",
  "message": "Full authentication is required to access this resource",
  "details": [
    "AuthenticationException"
  ],
  "errorCode": 11,
  "argumentsList": null
}

**Comment by MartinWahnschaffe** [comment]
@cazacmarin In sormas-rest we are so far using the ExceptionMapper of jax-rs (https://dennis-xlc.gitbooks.io/restful-java-with-jax-rs-2-0-2rd-edition/content/en/part1/chapter7/exception_handling.html). My first thought on this is that it would make more sense to go on with that approach instead of introducing a spring mechanism.
Do you think it would be better to go with the approach you have described above? If so, why?

A small notice on the I18n stuff: I'd say we should use the existing approach based on I18nProperties. Maybe use an enum for the different error codes you mentioned.

**Comment by cazacmarin** [comment]
@MartinWahnschaffe 
As investigation took a bit longer, I will put as well 1st  part of the investigation, which cannot be implemented, and 2nd one, which can.
---------------------------------------------------------------
Part 1-only to see how the investigation was.
---------------------------------------------------------------

I did some tests, in order to have a clean answer to right approach.
General idea is to create builders. And those builders, will populate data from exception accordingly to error code and other details at the moment when exception is thrown.
In order to implement it, is needed that all controllers will extend from `CustomizedExceptionHandler` class witch should extend `org.springframework.web.servlet.mvc.method.annotation.ResponseEntityExceptionHandler` class. 
1st and main problem which I saw: if, we are extending from `javax.ws.rs.WebApplicationException` in the manner which I described, call to the controller (any methods) will lead to 500 server error with general exception of injection. This one I cannot bypass in any manner.

2nd important  problem is the generated output. We definitely need to replace `javax.ws.rs.core.Response` to `org.springframework.http.ResponseEntity`. Content generated by jax-rs library, contain much more details than needed. And builder, will be represented in `toString` format within section `reason_phrase`. It means that client side need to parse this exception, to extract from it details from `toString` method which can change in time. (not a good practice). 
Output generated files are the next one:
[response-with-javax.ws.rs.core.Response.json.log](https://github.com/hzi-braunschweig/SORMAS-Project/files/8618036/response-with-javax.ws.rs.core.Response.json.log)
[response-with-org.springframework.http.ResponseEntity.json.log](https://github.com/hzi-braunschweig/SORMAS-Project/files/8618037/response-with-org.springframework.http.ResponseEntity.json.log)

3rd At mapping exception with jax-rs, sometime we can have mapping exception. And only general message will be thrown. Builder message itself, will not be created. In order to fix it, additional library should be added(and this one needs to be tested to be sure that output exception was created by builder):
```
<dependency>
            <groupId>org.glassfish.jersey.core</groupId>
            <artifactId>jersey-common</artifactId>
            <version>2.22.2</version>
            <scope>compile</scope>
        </dependency>
```
Based on that, I can conclude that we need to proceed with spring library.

---------------------------------------------------------------
Part 2 - real implementation
---------------------------------------------------------------
Due to the way how `sormas-rest` is created and Swagger app how is configured and registering controllers, spring library cannot be used.
Finally, I was able to do it with jax-rs. After a lot of issues, solution is the next one:
1. `sormas-api`   shared module will be used as an intermediate for accessing `sormas-rest` and `sormas-backend` project.
2.   `sormas-api`  will contain entire custom  `CustomizedException`  mechanism.
3. Customized error message can be send from both modules ( `sormas-rest` and `sormas-backend`)

I created the patch as a proof of concept and added it here: (annotations and inheritance as is, are very important to be there, otherwise, something will not work.
[#8934_-_Build_custom_Error_Details_message_for_REST_API.zip](https://github.com/hzi-braunschweig/SORMAS-Project/files/8652233/8934_-_Build_custom_Error_Details_message_for_REST_API.zip)

Can be used as well translation from server side. 

Output, looks like next one:
![image](https://user-images.githubusercontent.com/6985315/167428540-6079a0eb-229a-4dc2-90b1-149fcbd25883.png)

ArgumetList  var ca be used to customize some exception.

Afterword, if all is agreed, module by module tickets ca be opened and this mechanism can be applied for entire `rest-api` module.

**Comment by cazacmarin** [comment]
(moving ticket into  wait column as Martin input is required on it. When we will agree that is is good to be implemented, all relevant tickets will be opened for rest APIs and this ticket will be closed.)

**Comment by MartinWahnschaffe** [comment]
Thanks for diving into this. I looked into your new proposal and the patch and have a few comments:

1. Could you create and publish a branch for the patch here, so we can add comments to specific aspects of the code more easily?
2. I don't get when we would need the ExceptionDto with the id and the arguments list. Do you have an example for this?
3. getAllUuids should no return a bad request, when no data is found. It's still a valid request - there is just no data in the system
4. In contrast it would make sense that getByUuid returns a 404
5. I think we should map those error messages to http status codes
6. Instead of using string constants in the ExceptionsTemplate class I'd suggest to use an enum class and thus also provide I18n
7. Where are those authentication related constants like TOKEN_IS_NOT_VALID_MSG used? We probably need an exception mapper for authentication related exceptions, right?

All in all I think what we are still missing here is to take a step back to get an overview and define the following:

1. Which exceptions are thrown from the backend and rest application?
2. Which exceptions are missing and should be thrown? E.g. *Resource.getByUuid could throw an "EntityNotFoundException" or maybe a similar exception that already exists.
3. For each of those exceptions we should define how they are mapped. Currently we only have a mapper for ConstraintViolationException and EJBAccessExceptionMapper
4. This mapping could rely on an enum that defines all the potential ReST error results, including HTTP status code and an I18n message.

@vidi42 @JonasCir I'd also like your input on this. Probably worth to setup a meeting with the 4 of us once you have had a look at this. Maybe also Fredrik?

**Comment by cazacmarin** [comment]
Hi @MartinWahnschaffe. Some answers to your questions:
- `2.I don't get when we would need the ExceptionDto with the id and the arguments list. Do you have an example for this?`
	[Marin]arguments list is used to pass some additional variable to front. In case of angular if translation will be on the front side of app.
  in case if translation will be only on backend side, than this list is not needed.
  - `3. getAllUuids should no return a bad request, when no data is found. It's still a valid request - there is just no data in the system
  - 4. In contrast it would make sense that getByUuid returns a 404.`   [Marin]   (still thinking which the solution is the best one 404 or 204...)Currently  implemented with 404.
  - `5. I think we should map those error messages to http status codes` [Marin]  - generally speaking, yes. But sometime can exist scenarios in which are required to send a specific error code. So in my opinion both options should be supported.
  - `6. Instead of using string constants in the ExceptionsTemplate class I'd suggest to use an enum class and thus also provide I18n` [Marin] - make sense, 2nd commit contain this implementation.
  - 7. `Where are those authentication related constants like TOKEN_IS_NOT_VALID_MSG used? We probably need an exception mapper for authentication related exceptions, right?`  [Marin]-  Yes, that's true. I will try to implement it as well.
  
   All in all I think what we are still missing here is to take a step back to get an overview and define the following:
   ` 1. Which exceptions are thrown from the backend and rest application?`
	 [Marin] - In my opinion is good if we can do all on  backend side. In case of error modiffication, will be much more easy if we will modify the exception only on backend side.
	 `2. Which exceptions are missing and should be thrown? E.g. *Resource.getByUuid could throw an "EntityNotFoundException" or maybe a similar exception that already exists.` 
	 [Marin] -   NEED to find all the list from them. 
	 `3. For each of those exceptions we should define how they are mapped. Currently we only have a mapper for ConstraintViolationException and EJBAccessExceptionMapper`  [Marin]  -  usually this can be progressively done 1 by 1 when this mechanism will be implemented and all relevant tickets will be opened
	 `4. This mapping could rely on an enum that defines all the potential ReST error results, including HTTP status code and an I18n message.`  [Marin] - Totally agreed. I will think about it's implementation as well.

**Comment by cazacmarin** [comment]
![image](https://user-images.githubusercontent.com/6985315/170937624-6bdeb154-84da-4af0-a188-9663a255ae19.png)
2nd commit error message. In case if additional fields are still required, please let here a note.

**Comment by FredrikSchaeferVitagroup** [comment]
Related to the [2. in the definition need pointed out](https://github.com/hzi-braunschweig/SORMAS-Project/issues/8934#issuecomment-1128600277) by @MartinWahnschaffe , I would like to make a start.

When pushing invalid entities, users sometimes only get a `400 Bad Request` returned. When you don't have access to the SORMAS server logs, this makes it very hard/impossible to see the cause of the problem. I think it would be very helpful, if instead an explanation of what's wrong with the pushed entity gets returned, so that the sender can find out what to fix.
For some scenarios, this is already done (e.g. when not sending some mandatory attributes).

When a sending user does not have sufficient rights, currently only something rather generic gets returned (like `403 Forbidden`). Maybe it would help to point out what right is missing? Like that, it would also be easier to distinguish user rights issues from server configuration issues, which might also lead to 403s.

**Comment by JonasCir** [comment]
In general I think we should get rid of our custom approach and stick with exception mappers. They are quite flexible and directly integrated into the stack.

**Comment by cazacmarin** [comment]
Based on `https://stackoverflow.com/questions/3293599/jax-rs-using-exception-mappers` - it is possible to map exception builders. 
@JonasCir - if you can, it is better to make a PoC as we will be able to compare which solution is better.
In link provided, there were a lot of if/else statement, and not a normal, parametrized, with different arguments and if needed with internationalization thrown error.

When you will have a nicely solution as was provided by custom approach which can easy extend and `have the same message format` for all of the exceptions, please leave here a note, and I will be glad to take a look on it :) 
Thanks!

**Comment by JonasCir** [comment]
We can get rid of the if-else chains as jax will go through the super classes of an exception to find a matching handler, so we can start generic and then specialize if necessary. I think we can enforce a common scheme. For example, there is https://www.rfc-editor.org/rfc/rfc7807 which could help. I guess there will be a library or we can define a response entity class of a reasonable format which we serialize.

**Comment by cazacmarin** [comment]
Agree. All is possible :) - it is matter of time for new investigation. Due to that, I asked to have a branch with `a new proposal`, and only after, we can raise all `pros` and `cons` of both solutions. And only after, we can decide which one is the best one. And `win` solution will be followed. Hope it make sense for everybody :)
